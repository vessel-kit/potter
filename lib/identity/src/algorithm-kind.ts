import { InvalidAlgorithmKindError } from "./invalid-algorithm-kind.error";

/**
 * Supported [JOSE algorithms](https://www.iana.org/assignments/jose/jose.xhtml#web-signature-encryption-algorithms).
 */
export enum AlgorithmKind {
  /**
   * `secp256k1` signing of `SHA-256` digest as specified in [RFC 8812](https://tools.ietf.org/html/rfc8812).
   */
  ES256K = "ES256K",
  /**
   * `ed25519` signing as specified in [RFC 8037](https://tools.ietf.org/html/rfc8037).
   */
  EdDSA = "EdDSA",
}

/* istanbul ignore next */
export namespace AlgorithmKind {
  /**
   * Convert unknown external `input` to [[AlgorithmKind]].
   */
  export function fromString(input: string): AlgorithmKind {
    const searching = input as AlgorithmKind;
    switch (searching) {
      case AlgorithmKind.EdDSA:
        return AlgorithmKind.EdDSA;
      case AlgorithmKind.ES256K:
        return AlgorithmKind.ES256K;
      default:
        throw new InvalidAlgorithmKindError(searching);
    }
  }
}
