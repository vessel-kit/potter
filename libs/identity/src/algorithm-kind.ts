import { InvalidAlgorithmKindError } from './invalid-algorithm-kind.error';

export enum AlgorithmKind {
  ES256K = 'ES256K',
  EdDSA = 'EdDSA',
}

/* istanbul ignore next */
export namespace AlgorithmKind {
  export function fromString(input: string) {
    const searching = input as AlgorithmKind;
    switch (searching) {
      case AlgorithmKind.EdDSA:
        return AlgorithmKind.EdDSA;
      case AlgorithmKind.ES256K:
        return AlgorithmKind.ES256K;
      default:
        throw new InvalidAlgorithmKindError(searching);
    }
  }
}
