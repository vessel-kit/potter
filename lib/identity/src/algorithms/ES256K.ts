import * as bytes from '@stablelib/bytes';
import * as elliptic from 'elliptic';
import * as sha256 from '@stablelib/sha256';
import { IPrivateKey, ISigner } from '../private-key.interface';
import { AlgorithmKind } from '../algorithm-kind';
import { IPublicKey, ISignatureVerification } from '../public-key.interface';
import BN from 'bn.js';

const secp256k1Context = new elliptic.ec('secp256k1');

export class PublicKey implements IPublicKey, ISignatureVerification {
  readonly alg = AlgorithmKind.ES256K;
  constructor(readonly material: Uint8Array) {}

  async verify(message: Uint8Array, signature: Uint8Array): Promise<boolean> {
    try {
      const keyPair = secp256k1Context.keyFromPublic(this.material);
      const r = new BN(signature.slice(0, 32));
      const s = new BN(signature.slice(32, 64));
      const digest = sha256.hash(message);
      return keyPair.verify(digest, { r, s });
    } catch {
      return false;
    }
  }
}

export class PrivateKey implements IPrivateKey, ISigner {
  readonly alg = AlgorithmKind.ES256K;

  #keyPair: elliptic.ec.KeyPair;
  #publicKey: Uint8Array;

  constructor(material: Uint8Array) {
    this.#keyPair = secp256k1Context.keyFromPrivate(material);
    this.#publicKey = new Uint8Array(this.#keyPair.getPublic().encodeCompressed());
  }

  async publicKey(): Promise<PublicKey> {
    return new PublicKey(this.#publicKey);
  }

  async sign(message: Uint8Array): Promise<Uint8Array> {
    const digest = sha256.hash(message);
    const signature = this.#keyPair.sign(digest, { canonical: true });
    const r = new Uint8Array(signature.r.toArray('be', 32));
    const s = new Uint8Array(signature.s.toArray('be', 32));
    return bytes.concat(r, s);
  }
}
