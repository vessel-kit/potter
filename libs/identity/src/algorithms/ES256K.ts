import * as bytes from '@stablelib/bytes';
import * as elliptic from 'elliptic';
import { IPrivateKey, ISigner } from '../private-key.interface';
import { AlgorithmKind } from '../algorithm-kind';
import { IPublicKey } from '../public-key.interface';
import BN from 'bn.js';

const secp256k1Context = new elliptic.ec('secp256k1');

export class PublicKey implements IPublicKey {
  readonly kind = AlgorithmKind.ES256K;
  constructor(readonly material: Uint8Array) {}
}

export class PrivateKey implements IPrivateKey, ISigner {
  readonly alg = AlgorithmKind.ES256K;

  #keyPair: elliptic.ec.KeyPair;
  #publicKey: Uint8Array;

  constructor(material: Uint8Array) {
    this.#keyPair = secp256k1Context.keyFromPrivate(material);
    this.#publicKey = new Uint8Array(this.#keyPair.getPublic().encodeCompressed());
  }

  async publicKey(): Promise<PublicKey> {
    return new PublicKey(this.#publicKey);
  }

  async sign(message: Uint8Array): Promise<Uint8Array> {
    const signature = this.#keyPair.sign(message, { canonical: true });
    const r = new Uint8Array(signature.r.toArray('be', 32));
    const s = new Uint8Array(signature.s.toArray('be', 32));
    return bytes.concat(r, s);
  }
}

export function verifySignature(publicKey: IPublicKey, message: Uint8Array, signature: Uint8Array): boolean {
  try {
    const keyPair = secp256k1Context.keyFromPublic(publicKey.material);
    const r = new BN(signature.slice(0, 32));
    const s = new BN(signature.slice(32, 64));
    return keyPair.verify(message, { r, s });
  } catch {
    return false;
  }
}
