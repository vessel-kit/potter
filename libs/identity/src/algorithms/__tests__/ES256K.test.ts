import * as secp256k1 from '../ES256K';
import * as _ from 'lodash';
import { AlgorithmKind } from '../../algorithm-kind';

describe('PublicKey', () => {
  test('fields', () => {
    const material = new Uint8Array(_.times(32, () => 1));
    const key = new secp256k1.PublicKey(material);
    expect(key.kind).toEqual(AlgorithmKind.ES256K);
    expect(key.material).toEqual(material);
  });
});

describe('PrivateKey', () => {
  const material = new Uint8Array(_.times(32, () => 1));
  const key = new secp256k1.PrivateKey(material);
  test('fields', async () => {
    expect(key.alg).toEqual(AlgorithmKind.ES256K);
    const publicKey = await key.publicKey();
    expect(publicKey).toMatchSnapshot();
  });
  test('sign', async () => {
    const signature = await key.sign(material);
    expect(signature).toMatchSnapshot();
  });
});

describe('verifySignature', () => {
  const material = new Uint8Array(_.times(32, () => 1));
  const key = new secp256k1.PrivateKey(material);
  const message = new Uint8Array(_.times(32, (n) => n));
  test('ok', async () => {
    const publicKey = await key.publicKey();
    const signature = await key.sign(message);
    expect(secp256k1.verifySignature(publicKey, message, signature)).toBeTruthy();
  });
  test('wrong input', async () => {
    const publicKey = await key.publicKey();
    const signature = await key.sign(message);
    expect(secp256k1.verifySignature(publicKey, new Uint8Array(), signature)).toBeFalsy();
  });
});
