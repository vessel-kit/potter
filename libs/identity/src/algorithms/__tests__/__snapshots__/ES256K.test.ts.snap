// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateKey fields 1`] = `
PublicKey {
  "alg": "ES256K",
  "material": Uint8Array [
    3,
    27,
    132,
    197,
    86,
    123,
    18,
    100,
    64,
    153,
    93,
    62,
    213,
    170,
    186,
    5,
    101,
    215,
    30,
    24,
    52,
    96,
    72,
    25,
    255,
    156,
    23,
    245,
    233,
    213,
    221,
    7,
    143,
  ],
}
`;

exports[`PrivateKey sign 1`] = `
Uint8Array [
  99,
  53,
  129,
  21,
  236,
  176,
  92,
  167,
  204,
  137,
  178,
  140,
  146,
  96,
  142,
  102,
  239,
  166,
  121,
  167,
  238,
  85,
  225,
  209,
  113,
  234,
  25,
  125,
  180,
  187,
  122,
  15,
  80,
  184,
  169,
  2,
  119,
  174,
  39,
  165,
  143,
  129,
  98,
  191,
  142,
  97,
  26,
  62,
  59,
  75,
  184,
  208,
  24,
  29,
  29,
  115,
  85,
  104,
  22,
  157,
  220,
  52,
  220,
  0,
]
`;
